README.md
<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
**Table of Contents**  *generated with [DocToc](https://github.com/thlorenz/doctoc)*

- [secimport](#secimport)
  - [The Tailor-Made Sandbox for Your Application](#the-tailor-made-sandbox-for-your-application)
    - [The problem](#the-problem)
    - [The solution](#the-solution)
  - [Installation](#installation)
  - [With Docker](#with-docker)
  - [Without Docker](#without-docker)
  - [Usage](#usage)
  - [Stop on violation](#stop-on-violation)
  - [Kill on violation](#kill-on-violation)
  - [Dynamic profiling - trace, build sandbox, run.](#dynamic-profiling---trace-build-sandbox-run)
  - [nsjail support (seccomp)](#nsjail-support-seccomp)
  - [Python API](#python-api)
  - [Docker](#docker)
  - [Examples](#examples)
  - [Contributing](#contributing)
  - [Roadmap](#roadmap)
  - [Changelog](#changelog)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

# secimport
[![Upload Python Package](https://github.com/avilum/secimport/actions/workflows/python-publish.yml/badge.svg)](https://github.com/avilum/secimport/actions/workflows/python-publish.yml)
![](https://img.shields.io/badge/Test_Coverage-90%-blue)


## The Tailor-Made Sandbox for Your Application
secimport is production-oriented sandbox toolkit.<br>
It traces your code, and runs an executable that allows only the same syscalls, per module.

### The problem
Traditional tools like seccomp or AppArmor enforce syscalls for the entire process.<br>Something like `allowed_syscalls=["read","openat","bind","write"]`, which is great, but not enough for python's attack surface.<br>

### The solution
`secimport` is able to trace which syscalls each module in your code uses (by package name).<br>
After tracing, secimport creates a JSON/YAML policy for your code. It compiles it into a high-performance eBPF instrumentation script (smaller then 512 bytes overall), that looks like <a>this</a>.
```
modules:
  requests:
    destructive: true     # when true, secimport will kill on vilation instead of logging.
    syscall_allowlist:
      - fchmod
      - getentropy
      - getpgrp
      - getrlimit
...
```
You can also use JSON instaead of YAML, and even confine builtin python modules.<br>
An example policy that uses logging, multiprocessing, os and filesystem will look approximately like this:
```
...
    "/workspace/Python-3.10.0/Lib/logging/__init__.py": [
        " clock_gettime",
        " getpid",
        " write"
    ],
    "/workspace/Python-3.10.0/Lib/multiprocessing/process.py": [
        " getcwd",
        " getpid",
        " getrandom"
    ],
    "/workspace/Python-3.10.0/Lib/multiprocessing/util.py": [
        " prlimit64"
    ],
    "/workspace/Python-3.10.0/Lib/os.py": [
        " read"
    ],
    "/workspace/Python-3.10.0/Lib/platform.py": [
        " uname"
    ],
    "/workspace/Python-3.10.0/Lib/posixpath.py": [
        " close",
        " fstat",
        " getcwd",
        " getdents64",
        " openat"
    ],
    "/workspace/Python-3.10.0/Lib/random.py": [
        " getrandom"
    ],
...
```
It was created by tracing the code. secimport automatically finds these per-module syscalls for you.

Finally, you convert this policy into an sandbox (eBPF instrumentation script) to run the python process in production. Running the sandbox will enfore python to obey any given policy.

`secimport` is great for...
- Preventing Code Execution: reduce the risk of supply chain attacks.
  - Trace the syscalls flow of your application at the user-space/os/kernel level and per module.
  - Run your application while enforcing syscalls per module.
    - Upon violation of the policy, it can log, stop, or kill the process.
- Protect yourself from RCE:
  - secimport makes 1day attacks less of an issue, because it prevents the code form running. If you are using a vulnerable package and someone exploited it, your policy will not allow this exploit's syscalls and it will be handled as you wish.
  - Avoid incidents like <a href="https://en.wikipedia.org/wiki/Log4Shell">log4shell</a>. A logging library requires very few syscalls, and it should never run command using fork, execve or spawn.
    - The syscalls that "fastapi", "numpy" or "requests" use are very different.
- Load AI Models from Insecure Sources
  - Models from unsafe source (huggingface, torch hub, and pytorch pickled models) can be limited to run only a set of syscalls. RCE like 'import os;os.system(...)' somewhere deep in the code will be catched by secimport.
- Minimal Performance Impact
  -  Has negligible performance impact and is production-ready thanks to eBPF. Check out the [Performance](https://github.com/avilum/secimport/wiki/Performance-Benchmarks) benchmarks.
- Trace which syscalls are called by each module in your code.
  - secimport uses USDT (Userland Statically Defined Tracing) together with kernel probes in the runtime using eBPF or dtrace instrumentation scripts.
## Installation
Tested on Ubuntu, Debian, Rocky (Linux x86/AMD/ARM) and MacOS in (x86/M1). If you run on MacOS you will need to <a href="https://github.com/avilum/secimport/blob/master/docs/MAC_OS_USERS.md">disable SIP for dtrace. </a>

## With Docker
For quicker evaluation, we recommend using the <a href="[docs/Docker](https://github.com/avilum/secimport/blob/master/docker/README.md)">Secimport Docker Image</a> instead of self-installing.<br>
- Build and run the Docker container with a custom kernel that matches your existing OS kernel version:
  ```
  cd docker/ && ./build.sh && ./run.sh
  ```
  A temporary container will be created, and you will be logged in as the root user.

## Without Docker
1. Install python with USDT probes by <a href="https://github.com/avilum/secimport/wiki/Installation#python-interpreter-requirements">configuring it with '--dtrace'</a>
2. Install one of the backends: <a href="https://github.com/avilum/secimport/wiki/Installation">eBPF or DTrace</a>.
3. Install secimport
  - Install from pypi
    - ```
      python3 -m pip install secimport
      ```
  - Install from source
    - ```
      git clone https://github.com/avilum/secimport.git && cd secimport
      python3 -m pip install poetry && python3 -m poetry install
      ```


## Usage
To sandbox your program using the CLI, start a bpftrace program that logs all the syscalls for all the modules in your application into a file with the secimport trace command. Once you have covered the logic you would like to sandbox, hit CTRL+C or CTRL+D, or wait for the program to finish. Then, build a sandbox from the trace using the secimport build command, and run the sandbox with the secimport run command.

```shell
NAME
    secimport - is a comprehensive toolkit designed to enable the tracing, construction, and execution of secure Python runtimes. It leverages USDT probes and eBPF/DTrace technologies to enhance the overall security measures.

SYNOPSIS
    secimport COMMAND

DESCRIPTION
    https://github.com/avilum/secimport/wiki/Command-Line-Usage

    WORKFLOW:
            1. secimport trace / secimport shell
            2. secimport build
            3. secimport run

    QUICKSTART:
            $ secimport interactive

    EXAMPLES:
        1. trace:
            $  secimport trace
            $  secimport trace -h
            $  secimport trace_pid 123
            $  secimport trace_pid -h
        2. build:
            # secimport build
            $ secimport build -h
        3. run:
            $  secimport run
            $  secimport run --entrypoint my_custom_main.py
            $  secimport run --entrypoint my_custom_main.py --stop_on_violation=true
            $  secimport run --entrypoint my_custom_main.py --kill_on_violation=true
            $  secimport run --sandbox_executable /path/to/my_sandbox.bt --pid 2884
            $  secimport run --sandbox_executable /path/to/my_sandbox.bt --sandbox_logfile my_log.log
            $  secimport run -h

COMMANDS
    COMMAND is one of the following:

     build
       Compiles a trace log (trace.log). Creates the sandbox executable (instrumentation script) for each supported backend It uses `create_profile_from_trace ...` and `sandbox_from_profile`.

     compile_sandbox_from_profile
       Generates a tailor-made sandbox that will enforce a given yaml profile/policy in runtime.

     interactive

     run
       Run a python process inside the sandbox.

     shell
       Alternative syntax for secimport "trace".

     trace
       Traces a python process using an entrypoint or interactive interpreter. It might require sudo privilleges on some hosts.

     trace_pid
       Traces a running process by pid. It might require sudo privilleges on some hosts.
```

## Stop on violation
```
root@1bc0531d91d0:/workspace# secimport run  --stop_on_violation=true
 >>> secimport run
[WARNING]: This sandbox will send SIGSTOP to the program upon violation.
 RUNNING SANDBOX... ['./sandbox.bt', '--unsafe', ' -c ', '/workspace/Python-3.10.0/python', 'STOP']
Attaching 4 probes...
Python 3.10.0 (default, Apr 28 2023, 11:32:40) [GCC 9.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import os
>>> os.system('ps')
[SECURITY PROFILE VIOLATED]: <stdin> called syscall 56 at depth 8022

^^^ STOPPING PROCESS 85918 DUE TO SYSCALL VIOLATION ^^^
		PROCESS 85918 STOPPED.
```

## Kill on violation
```
root@ee4bc99bb011:/workspace# secimport run --kill_on_violation
 >>> secimport run
[WARNING]: This sandbox will send SIGKILL to the program upon violation.
 RUNNING SANDBOX... ['./sandbox.bt', '--unsafe', ' -c ', '/workspace/Python-3.10.0/python', 'KILL']
import os
oAttaching 4 probes...
sPython 3.10.0 (default, Apr 28 2023, 11:32:40) [GCC 9.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import os
>>> os.system('ps')
[SECURITY PROFILE VIOLATED]: <stdin> called syscall 56 at depth 8022

^^^ KILLING PROCESS 86466 DUE TO SYSCALL VIOLATION ^^^
		KILLED.
 SANDBOX EXITED;
```

## Dynamic profiling - trace, build sandbox, run.
```shell
root@1fa3d6f09989:/workspace# secimport interactive

Let's create our first tailor-made sandbox with secimport!
- A python shell will be opened
- The behavior will be recorded.

OK? (y): y
 >>> secimport trace

TRACING: ['/workspace/secimport/profiles/trace.bt', '-c', '/workspace/Python-3.10.0/python', '-o', 'trace.log']

                        Press CTRL+D to stop the trace;

Python 3.10.0 (default, Mar 19 2023, 08:34:46) [GCC 9.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import this
>>>
 TRACING DONE;
 >>> secimport build

SECIMPORT COMPILING...

CREATED JSON TEMPLATE:  policy.json
CREATED YAML TEMPLATE:  policy.yaml
compiling template policy.yaml
DTRACE SANDBOX:  sandbox.d
BPFTRCE SANDBOX:  sandbox.bt
```

Now, let's run the sandbox!
```python
- Run the same commands as before, they should run without any problem;.
- Do something new in the shell; e.g:   >>> __import__("os").system("ps")

        OK? (y): y
 >>> secimport run
 RUNNING SANDBOX... ['./sandbox.bt', '--unsafe', ' -c ', '/workspace/Python-3.10.0/python']
Attaching 5 probes...
REGISTERING SYSCALLS...
STARTED
Python 3.10.0 (default, Mar 19 2023, 08:34:46) [GCC 9.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import this
>>> import os
[SECIMPORT VIOLATION]: <stdin> called syscall ioctl at depth 0
[SECIMPORT VIOLATION]: <stdin> called syscall ioctl at depth 0
```

For more detailed usage instructions, see the [Command-Line Usage](https://github.com/avilum/secimport/wiki/Command-Line-Usage) page.

## nsjail support (seccomp)
Beside the sandbox that secimport builds, <br>
The `secimport build` command creates an <a href="https://github.com/google/nsjail">nsjail</a> sandbox with seccomp profile for your traced code.<br> `nsjail` enables namespace sandboxing with seccomp on linux<br>
`secimport` automatically generates seccomp profiles to use with `nsjail` as executable bash script.
It can be used to limit the syscalls of the entire python process, as another layer of defence.

## Python API

Instead of CLI, you can also use `secimport` by replacing "`import`" with "`secimport.secure_import`" for selected modules. See the [Python Imports](examples/python_imports/) example for more details.

## Docker
The quickest way to evaluate `secimport` is to use our [Docker container](docker/README.md), which includes `bpftrace` (`ebpf`) and other plug-and-play examples.


## Examples

The [Sandbox Examples](https://github.com/avilum/secimport/wiki/Sandbox-Examples) page contains basic and advanced real-world examples.

## Contributing

For information on how to contribute to `secimport`, see the [Contributing](https://github.com/avilum/secimport/blob/master/docs/CONTRIBUTING.md) guide.

## Roadmap

See the [Roadmap](https://github.com/avilum/secimport/blob/master/docs/ROADMAP.md) for the planned features and development milestones.

## Changelog

See the [Changelog](https://github.com/avilum/secimport/blob/master/docs/CHANGELOG.md) for development progress and existing features.


docs/CHANGELOG.md
<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
**Table of Contents**  *generated with [DocToc](https://github.com/thlorenz/doctoc)*

- [Changelog](#changelog)
  - [References](#references)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

# Changelog
- ✔️ Added Allow/Block list configuration
- ✔️ Created a .yaml configuration per module in the code
  - ✔️ Use secimport to compile that yml
  - ✔️ Create a single dcript policy
  - ✔️ Run an application with that policy using dtrace, without using `secure_import`
- ✔️ Added eBPF basic support using bpftrace
  - ✔️ bpftrace backend tests
- ✔️ Implemented python USDT probes template
- ✔️ Added CLI for bpftrace backend usage
- ✔️ Updated documentation and improved CLI
- ✔️ Added GIFs

## References
- Read more about the primitives of secimport:
  - `bpftrace` - https://github.com/iovisor/bpftrace
  - `dtrace` - [DTrace Cheatsheet](https://www.brendangregg.com/DTrace/DTrace-cheatsheet.pdf)
    - [DTrace for Linux (2018)](https://www.brendangregg.com/blog/2018-10-08/dtrace-for-linux-2018.html)
- <a href="https://github.com/avilum/secimport/wiki/Sandbox-Examples">Sandbox Examples</a>
- Guides
  - <a href="https://github.com/avilum/secimport/wiki/Tracing-Processes">Tracing Processes Guide</a>
  - <a href="https://github.com/avilum/secimport/wiki/Installation">Installation</a>
  - <a href="https://github.com/avilum/secimport/wiki/YAML-Profiles">Create a Sandbox from YAML file</a>
  - <a href="https://github.com/avilum/secimport/wiki/MacOS-Users">Mac OS Users</a>
  - <a href="https://github.com/avilum/secimport/wiki/F.A.Q">F.A.Q</a>


secimport/cli.py
from collections import defaultdict
import json
import os
from pathlib import Path
import stat
import fire
import sys
import yaml
import secimport
from secimport.backends.common.utils import SECIMPORT_ROOT


class COLORS:
    HEADER = "\033[95m"
    OKBLUE = "\033[94m"
    OKCYAN = "\033[96m"
    OKGREEN = "\033[92m"
    WARNING = "\033[93m"
    FAIL = "\033[91m"
    ENDC = "\033[0m"
    BOLD = "\033[1m"
    UNDERLINE = "\033[4m"


def colored_print(color, *args):
    print(color, *args, COLORS.ENDC)


class SecImportCLI:
    """
    secimport is a comprehensive toolkit designed to enable the tracing, construction, and execution of secure Python runtimes. It leverages USDT probes and eBPF/DTrace technologies to enhance the overall security measures.\n
    https://github.com/avilum/secimport/wiki/Command-Line-Usage

    WORKFLOW:
            1. secimport trace / secimport shell
            2. secimport build
            3. secimport run

    QUICKSTART:
            $ secimport interactive

    EXAMPLES:
        1. trace:
            $  secimport trace
            $  secimport trace -h
            $  secimport trace_pid 123
            $  secimport trace_pid -h
        2. build:
            # secimport build
            $ secimport build -h
        3. run:
            $  secimport run
            $  secimport run --entrypoint my_custom_main.py
            $  secimport run --entrypoint my_custom_main.py --stop_on_violation=true
            $  secimport run --entrypoint my_custom_main.py --kill_on_violation=true
            $  secimport run --sandbox_executable /path/to/my_sandbox.bt --pid 2884
            $  secimport run --sandbox_executable /path/to/my_sandbox.bt --sandbox_logfile my_log.log
            $  secimport run -h
    """

    @staticmethod
    def __create_profile_from_trace(
        trace_file: str = "trace.log",
        output_json_file: str = "policy.json",
        output_yaml_file: str = "policy.yaml",
        output_syscalls_file: str = "syscalls.txt",
        nsjail_sandbox_file: str = "nsjail-seccomp-sandbox.sh",
        system_interpreter: str = sys.executable,
        destructive: bool = True,
    ) -> str:
        """Creates a yaml policy file from a given trace log. default: "trace.log".

        After running this command, the yaml output file can be compiled into sandbox code using the 'sandbox_from_profile' command.
        """
        print(COLORS.WARNING)
        if not Path(trace_file).exists:
            raise FileNotFoundError(trace_file)

        parse_cmd = f"""cat {trace_file} | grep -Po "\@modules_syscalls\[\K(.+?)]" | tr -d ']' | sort > sandbox.log"""
        os.system(parse_cmd)
        modules_syscalls_file = Path("sandbox.log")
        if not modules_syscalls_file.exists():
            raise FileNotFoundError(trace_file)

        with open(modules_syscalls_file, "r") as f:
            lines = f.read().splitlines()

        modules_and_syscalls = defaultdict(list)
        syscalls = set()
        for line in lines:
            parsed_items = line.split(",")
            if parsed_items == 1:
                # General python import, interpreter dependency
                modules_and_syscalls["general_requirements"].append(parsed_items[0])
            elif len(parsed_items) == 2:
                # Module dependency
                _module, _syscall = parsed_items
                syscalls.add(_syscall)
                if not _module or _module == "None":  # empty string and None
                    _module = "general_requirements"
                modules_and_syscalls[_module].append(_syscall)
            else:
                raise Exception("there are more than 2 items in one of the lines.")

        # Creating a syscalls file
        _syscalls = list(sorted(list(syscalls)))
        syscalls_txt = ", ".join(_syscalls)
        nsjail_sandbox_cmd = f"nsjail -Ml -Mo  --chroot / --port 8000 --user 99999 --group 99999 --seccomp_string 'ALLOW {{ {syscalls_txt} }} DEFAULT KILL' -- {system_interpreter} -i"
        print(f"syscalls: {syscalls_txt}")
        print(f"nsjail sandbox: {nsjail_sandbox_cmd}")

        # Sandbox
        with open(output_syscalls_file, "w") as f:
            f.write(syscalls_txt)

        # NSJail
        with open(nsjail_sandbox_file, "w") as f:
            f.write(nsjail_sandbox_cmd)
        st = os.stat(nsjail_sandbox_file)
        os.chmod(nsjail_sandbox_file, st.st_mode | stat.S_IEXEC)

        # JSON
        modules_json = json.dumps(modules_and_syscalls, indent=4)
        with open(output_json_file, "w") as f:
            f.write(modules_json)

        # Building the YAML template
        yaml_dict = {
            name: {
                "destructive": destructive,
                "syscall_allowlist": _syscalls,
            }
            for name, _syscalls in modules_and_syscalls.items()
        }
        yaml_dict = {"modules": yaml_dict}
        modules_yaml = yaml.safe_load(json.dumps(yaml_dict))
        modules_yaml = yaml.dump(modules_yaml)
        with open(output_yaml_file, "w") as f:
            f.write(modules_yaml)
        print("Policy is ready: ", output_yaml_file, output_json_file)

        colored_print(COLORS.ENDC)
        return output_yaml_file, output_syscalls_file, nsjail_sandbox_file

    @staticmethod
    def compile_sandbox_from_profile(
        template_filename: str = "policy.yaml", sandbox_filename_prefix: str = "sandbox"
    ) -> str:
        """Generates a tailor-made sandbox that will enforce a given yaml profile/policy in runtime.

        2 scripts will be created in the same directory: one for the dscript backend and one for the bpftrace backend.
        Returns: The file name of the bpftrace sandbox (preferred).
            _type_: str
        """
        print(COLORS.WARNING)
        template_filename = Path(template_filename)
        dtrace_sandbox_filename = f"{sandbox_filename_prefix}.d"
        dtrace_rendered_profile = secimport.build_module_sandbox_from_yaml_template(
            template_path=template_filename,
            backend=secimport.InstrumentationBackend.DTRACE,
        )
        if not dtrace_rendered_profile:
            # empty file, error is already printed to the user
            raise AssertionError(
                f"The trace file '{template_filename}' has 0 modules. A valid policy should include at least 1 module."
            )

        with open(dtrace_sandbox_filename, "w") as f:
            f.write(dtrace_rendered_profile)

        st = os.stat(dtrace_sandbox_filename)
        os.chmod(dtrace_sandbox_filename, st.st_mode | stat.S_IEXEC)
        print("DTRACE sandbox:", dtrace_sandbox_filename)

        bpftrace_sandbox_filename = f"{sandbox_filename_prefix}.bt"
        bpftrace_rendered_profile = secimport.build_module_sandbox_from_yaml_template(
            template_path=template_filename,
            backend=secimport.InstrumentationBackend.EBPF,
        )

        with open(bpftrace_sandbox_filename, "w") as f:
            f.write(bpftrace_rendered_profile)
        st = os.stat(bpftrace_sandbox_filename)
        os.chmod(bpftrace_sandbox_filename, st.st_mode | stat.S_IEXEC)

        colored_print(COLORS.ENDC)
        return bpftrace_sandbox_filename

    @staticmethod
    def shell(*args, **kwargs):
        """Alternative syntax for secimport "trace"."""
        SecImportCLI.trace(*args, **kwargs)

    @staticmethod
    def trace(
        entrypoint: str = None,
        python_interpreter: str = sys.executable,
        trace_log_file: str = "trace.log",
    ):
        """Traces a python process using an entrypoint or interactive interpreter. It might require sudo privilleges on some hosts.
        Args:
            entrypoint (str, optional): A python script path to trace. If not specified, a python interactive shell will be opened in the foregraound. This shell, or the given command, will be traced and logged to the "trace_log_file" argument.
            python_interpreter (str, optional): The path of the python executable interpreter to trace. Defaults to sys.executable.
            trace_log_file (str, optional): The log file to write the trace into. Defaults to "trace.log".
        """
        if entrypoint:
            # entrypoint_cmd = str(Path(entrypoint).absolute())
            entrypoint_cmd = f'bash -c "{python_interpreter} {entrypoint}"'
        else:
            entrypoint_cmd = python_interpreter
        cmd = [
            f"{SECIMPORT_ROOT}/profiles/trace.bt",
            "-c",
            f"{entrypoint_cmd}",
            "-o",
            f"{trace_log_file}",
        ]
        colored_print(COLORS.HEADER, "\nTracing using ", cmd)
        colored_print(
            COLORS.BOLD,
            "Press CTRL+D or CTRL+C to stop the trace gracefully.\n",
        )
        os.system(" ".join(cmd))
        colored_print(
            COLORS.BOLD, "The trace log is at", COLORS.OKGREEN, f"./{trace_log_file}"
        )

    @staticmethod
    def trace_pid(
        pid: int,
        trace_log_file: str = "trace.log",
    ):
        """Traces a running process by pid. It might require sudo privilleges on some hosts.

        Args:
            pid (int): process pid to trace
            trace_log_file (str): The output file to write the trace into.
        """
        assert isinstance(pid, int), f"pid must be an int, got: {pid}"
        cmd = [
            str(SECIMPORT_ROOT) + "/profiles/trace.bt",
            "-p",
            f"{pid}",
            "-o",
            f"{trace_log_file}",
        ]
        colored_print(COLORS.HEADER, "\nTRACING PID:")
        colored_print(COLORS.OKBLUE, pid)
        input("\t\t\tPress CTRL+D/CTRL+C to stop the trace;")
        os.system(" ".join(cmd))
        colored_print(COLORS.ENDC)

    @staticmethod
    def build(
        trace_file: str = "trace.log",
        output_json_file: str = "policy.json",
        output_yaml_file: str = "policy.yaml",
        output_syscalls_file: str = "syscalls.txt",
        destructive: bool = False,
    ) -> str:
        """Compiles a trace log (trace.log). Creates the sandbox executable (instrumentation script) for each supported backend
        It uses `create_profile_from_trace ...` and `sandbox_from_profile`.

        Raises:
            FileNotFoundError:

        Returns:
            _type_: str
        """
        (
            output_yaml_file,
            output_syscalls_file,
            nsjail_sandbox_file,
        ) = SecImportCLI.__create_profile_from_trace(
            trace_file=trace_file,
            output_json_file=output_json_file,
            output_yaml_file=output_yaml_file,
            output_syscalls_file=output_syscalls_file,
            destructive=destructive,
        )

        bpftrace_sandbox_filename = SecImportCLI.compile_sandbox_from_profile(
            template_filename=output_yaml_file
        )
        colored_print(COLORS.OKGREEN, f"syscalls: {COLORS.BOLD} {output_syscalls_file}")
        colored_print(
            COLORS.OKGREEN, f"nsjail sandbox: {COLORS.BOLD} {nsjail_sandbox_file}"
        )
        colored_print(
            COLORS.OKGREEN,
            f"secimport eBPF sandbox: {COLORS.BOLD} {bpftrace_sandbox_filename}",
        )

    @staticmethod
    def run(
        entrypoint: str = None,
        sandbox_executable: str = "./sandbox.bt",
        python_interpreter: str = sys.executable,
        pid=None,
        stop_on_violation=False,
        kill_on_violation=False,
        sandbox_logfile: str = None,
    ):
        """Run a python process inside the sandbox.

        Args:
            entrypoint (str, optional): A script entrypoint to trace explicitly. Defaults to None.
            sandbox_executable (str, optional): An executable sandbox file, generated by "secimport build". Defaults to "./sandbox.bt".
            python_interpreter (str, optional): The python interpreter to probe. Defaults to sys.executable.
            pid (_type_, optional): process id to trace. Defaults to None.
            stop_on_violation: (str, optional). If set to True, the sandbox will send SIGSTOP to the process and prevent the execution the syscall.
            kill_on_violation: (str, optional). If set to True, the sandbox will send SIGKILL to the process, and then the sandbox will exit.
            sandbox_logfile: (str, optional). Log the sandbox STDOUT to an explicit file, instead of the foreground STDOUT/STDERR.
        Raises:
            FileNotFoundError: _description_
            ValueError: _description_
        """
        if not Path(sandbox_executable).exists:
            raise FileNotFoundError(
                "The sandbox was not found. Try calling 'secimport trace/trace_pid' and then 'secimport compile_sandbox'"
            )
        sandbox_startup_cmd = [f"{sandbox_executable}", "--unsafe"]
        if pid is not None:
            sandbox_startup_cmd += [f" -p {pid}"]
        elif python_interpreter is not None:
            if entrypoint:
                sandbox_startup_cmd += [
                    " -c ",
                    f'bash -c "{python_interpreter} {entrypoint}"',
                ]
            else:
                sandbox_startup_cmd += [" -c ", f"{python_interpreter}"]
        else:
            raise ValueError("Please specify one of the following: entrypoint / pid")
        if sandbox_logfile:
            sandbox_startup_cmd += [" -o sandbox.log"]

        # Passing the policy as bpftrace program arguments (They used as $1, $2, ...)
        assert not (
            stop_on_violation and kill_on_violation
        ), "Please specify only one of the following flags: `--stop_on_violation`, `--kill_on_violation`"
        if stop_on_violation:
            print(
                "[WARNING]: This sandbox will send SIGSTOP to the program upon violation."
            )
            sandbox_startup_cmd.append("STOP")
        elif kill_on_violation:
            print(
                "[WARNING]: This sandbox will send SIGKILL to the program upon violation."
            )
            sandbox_startup_cmd.append("KILL")

        colored_print(COLORS.HEADER, "RUNNING SANDBOX...", sandbox_startup_cmd)
        os.system(" ".join(sandbox_startup_cmd))
        colored_print(
            COLORS.BOLD,
            "SANDBOX EXITED;",
            f"Log file: {sandbox_logfile};" if sandbox_logfile else "",
        )

    @staticmethod
    def interactive():
        colored_print(COLORS.BOLD, "Welcome to secimport!")
        colored_print(
            COLORS.OKBLUE,
            "1. A python code can be traced using",
            COLORS.OKGREEN,
            '"secimport trace"',
            COLORS.OKBLUE,
            "or using",
            COLORS.OKGREEN,
            '"secimport shell"',
            COLORS.OKBLUE,
            "\n 2. All the function calls and system calls will be recorded to",
            COLORS.OKGREEN,
            "./trace.log.\n",
        )
        start = input("Continue? (y): ") or "y"
        if start.lower() == "y":
            try:
                SecImportCLI.trace()
            except KeyboardInterrupt:
                ...

            colored_print(
                COLORS.OKBLUE,
                '\nRunning "secimport build"; it compiles ./trace.log which is the trace log to ./sandbox.bt',
            )
            SecImportCLI.build()

            colored_print(
                COLORS.OKBLUE,
                "\nNow, let's run the sandbox.\n1. Run the same commands as before. They should run without any problem.\n2. Do something new in the shell. For example, ",
                COLORS.BOLD,
                "run \"import os; os.system('ps'); and see how secimport detects it.\"",
            )
            run = input("\nContinue? (y): ") or "y"

            if run.lower() == "y":
                SecImportCLI.run()
        else:
            colored_print(
                COLORS.OKBLUE,
                run,
                "Please type 'y' to start and follow the instructions; The docs are at https://github.com/avilum/secimport/wiki/Command-Line-Usage",
            )
        colored_print(COLORS.ENDC)


def main():
    import atexit

    atexit.register(lambda: colored_print(COLORS.ENDC))
    fire.Fire(SecImportCLI())


if __name__ == "__main__":
    main()
